omen2/omen.py

```diff
@@ -0,0 +1,214 @@
+
+    def __init_subclass__(cls, **_kws):
+        cls.table_types = {}
+
+        if not cls.model:
+            db = SqliteDb(":memory:")
+            cls.__multi_query(db, cls.schema(cls.version))
+            cls.model: DbModel = db.model()
+
+    def __init__(self, db: DbBase, migrate=True, **table_types):
+        """Create a new manager with a db connection."""
+        self.tables = {}
+        self.db = db
+
+        if migrate:
+            self._migrate_if_needed()
+        self._create_if_needed()
+
+        self.table_types.update(table_types)
+
+        for name, table_type in self.table_types.items():
+            # allow user to specify the table name this way instead
+            if not hasattr(table_type, "table_name"):
+                table_type.table_name = name
+            table_type(self)
+
+    def get_table_by_name(self, table_name):
+        return self[self.table_types[table_name]]
+
+    def __getitem__(self, table_type: Type[T]) -> T:
+        return self.tables[table_type]
+
+    def __setitem__(self, table_type: Type[T], table: T):
+        assert table_type.table_name == table.table_name
+        self.table_types[table.table_name] = table_type
+        self.validate_table(table.table_name, table_type)
+        self.tables[table_type] = table
+
+    def set_table(self, table: Table):
+        self[type(table)] = table
+
+    def load_dict(self, data_set: Dict[str, Iterable[Dict[str, Any]]]):
+        # load sample data into self
+        for name, values in data_set.items():
+            tab: Table = self.get_table_by_name(name)
+            for entry in values:
+                tab.add(tab.row_type._from_db(entry))
+
+    def dump_dict(self) -> Dict[str, Iterable[Dict[str, Any]]]:
+        ret = {}
+        for ttype, tab in self.tables.items():
+            lst = []
+            for obj in tab:
+                lst.append(obj._to_db())
+            ret[ttype.table_name] = lst
+        return ret
+
+    def validate_table(self, name, tab):
+        assert issubclass(tab, Table)
+
+        if not getattr(tab, "table_name", None):
+            tab.table_name = name
+        assert tab.table_name == name
+
+        assert issubclass(tab.row_type, ObjBase)
+        assert isinstance(getattr(tab.row_type, "_pk"), tuple)
+        assert issubclass(tab.row_type._table_type, tab)
+
+        if not getattr(tab, "field_names", None):
+            log.debug("%s: default serialization field names used", name)
+            tab.field_names = {c.name for c in self.model[name].columns}
+        assert isinstance(tab.field_names, set)
+
+        pk = None
+        model = self.model[name]
+        for idx in model.indexes:
+            if idx.primary:
+                pk = idx.fields
+        if len(pk) == 1:
+            for fd in model.columns:
+                if fd.name == pk[0]:
+                    tab.allow_auto = fd.autoinc
+
+    @classmethod
+    def codegen(cls, force=False):
```

Comment:
why is there a force command?   and please add a path variable.   ok for there to be an only_if_missing=False variable, and ok for path to be optional